### PR TITLE
Updated interface to throttled

### PR DESCRIPTION
**Pull request by metcalf**
This implements a substantially updated interface to throttled. The primary goals are:
* Reduce the overall surface area of the package to ease consumption and maintenance.
* Permit using the rate limiter without an http.Handler. This makes it more composable and allows for use in cases where e.g. you want to do some input validation on a request before running rate limiting.
* Switch to sliding window rate limiting using a modified generic cell rate algorithm.
* Maintain compatibility with what I would suspect is the most common production usage of the package:
```
st := store.NewRedisStore(myPool, "pfx-", 0)
vary := &VaryBy{RemoteAddr: true}
t := throttled.RateLimit(PerMin(5), vary, st)
throttledHandler := t.Throttle(myHandler)
```

For more context, see the discussion on #8.

TODO:
* [x] Get Travis CI running and passing (depends on becoming an organization owner)
* [x] Update/write README and doc.go
* [x] Update/write examples
* [x] Experiment with a `GCRAStore` interface that's more time aware. (nope)
* [x] Possibly expose the store testing helpers in the public interface for others to use. (possibly in the future)
* [x] Possibly accept a quantity parameter to Limit to allow for limiting of variable quantities. (possibly in the future)
* [x] Consider retaining `Q` and `Quota` instead of introducing `RateQuota`. (nope)
* [x] Cut a 1.0 release from existing master and update imports accordingly.

### CODE REVIEW

**Comment by metcalf** [comment]
@brandur: can you add me to the organization so I can get Travis set up? Thanks!

**Comment by brandur** [comment]
@metcalf Done!

**Comment by metcalf** [comment]
@brandur I'd be interested in your input on a few thoughts that came up during the implementation:
* There's a bunch of ugly code in `redisStore` to deal with pre-2.6 Redis servers that don't support `EVAL`. The last 2.4.x release was 3 years ago -- not crazy old but maybe old enough to consider excluding support,
* The definition of `RateQuota` isn't functionally different from the existing `Q` type and `Quota` interface. I don't love the naming (nor `Q.Requests` and `Q.Window`) but I'm not sure it's worth deprecating them until they're actually confusing.
* `Limit` could accept a 2nd parameter representing the quantity by which to increment the rate limit counter. This would allow, for example, efficiently limiting the total throughput of an interface.
```
type Limiter interface {
	Limit(key string, count int) (LimitResult, error)
}
```
* The TTL parameter to the GCRA store functions are always exactly equal to the value we're trying to set. I also spend a bit of time marshaling between `int64` and `time.Time` in the rate limiting code. That feels kind of repetitive for an already unusual interface but might be easier to grok. The interface below is less repetitive but feels a little too weird for me. Any other ideas?
```
// GCRAStore is the interface to implement to store state for a GCRA rate limiter
type GCRAStore interface {
	// Get returns the value of the key if it is in the store or a zero-value if it does
	// not exist. Also returns the current time at the Store. 
	GetTimeWithNow(key string) (time.Time, time.Time, error)

	// SetTimeIfNotExists sets the value of key only if it is not already set in the store
	// it returns whether a new value was set. If the store supports expiring
	// keys and a new value was set, the key will expire after its time value passes.
	SetTimeIfNotExists(key string, value time.Time) (bool, error)

	// CompareAndSwap atomically compares the value at key to the old value.
	// If it matches, it sets it to the new value and returns true. Otherwise,
	// it returns false. If the key does not exist in the store, it returns
	// false with no error. If the store supports expiring keys and the swap
	// succeeded, the key will expire after the new time value passes.
	CompareAndSwapTime(key string, old, new time.Time) (bool, error)
}
```

**Comment by brandur** [comment]
@metcalf

> There's a bunch of ugly code in redisStore to deal with pre-2.6 Redis servers that don't support EVAL. The last 2.4.x release was 3 years ago -- not crazy old but maybe old enough to consider excluding support,

+1 to stripping it out as it seems like three years is plenty of lead time.

Your other option I guess is maybe to just leave it in for now if it's cheap to do so, and get a separate pull started as a later phase in an effort to solicit specific feedback from anyone who may care. (But I wouldn't worry too much about doing this.)

> The definition of RateQuota isn't functionally different from the existing Q type and Quota interface. I don't love the naming (nor Q.Requests and Q.Window) but I'm not sure it's worth deprecating them until they're actually confusing.

On the other hand, if there's one time to get this fixed up, it's now. Upgrades should be pretty easy given that you'd made sure to keep the various helpers in place. Personally, I'd stay the course.

> Limit could accept a 2nd parameter representing the quantity by which to increment the rate limit counter. This would allow, for example, efficiently limiting the total throughput of an interface.

For this one, you mean that we'd calculate the new theoretical arrival time based off that that count right? Sort of like this:

``` go
if now.After(tat) {
    newTat = now.Add(count * g.emissionInterval)
} else {
    newTat = tat.Add(count * g.emissionInterval)
}
```

If so, that sounds great! It slipped my mind before, but one case we need to handle is where we actually vary fill rates even inside of the same limit based off of some out-of-band characteristic of the bucket itself. For example, we'll assign more generous rate limits for some of our larger customers. We call it a "multiplier" internally. This alternative interface would allow us to re-implement that feature on top of GCRA.

Possibly one thing to consider here is to provide two different interface functions for convenience so that the common case doesn't have to think about what the extra parameter might be for. Totally up to you though.

``` go
type Limiter interface {
    Limit(key string) (LimitResult, error)

    LimitWithCount(key string, count int) (LimitResult, error)
}
```

> The TTL parameter to the GCRA store functions are always exactly equal to the value we're trying to set. I also spend a bit of time marshaling between int64 and time.Time in the rate limiting code. That feels kind of repetitive for an already unusual interface but might be easier to grok. The interface below is less repetitive but feels a little too weird for me. Any other ideas?

I'm less sure on this one. I guess it's really about how much of the limiting logic you want to farm out to to your store. Theoretically you could have it all in there and your interface could just look the same as your `Limiter` interface:


``` go
type GCRAStore interface {
    Limit(key string) (LimitResult, error)
}
```

Generally, I'd be predisposed to err on the side of keeping things inside the application code, but you're right about `ttl` in that it does seem to be invariant and I can't think of a good reason that you'd ever want to make it anything other than `newTAT - now()`. I might still be tempted to keep it in the GRCA's core implementation just for clarity, but I'd +1 either approach.

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
+  - redis-server
 
-script: go test -v -short ./...
+install:
+  - make get-deps
+  # Move to the gopkg location that rather than the default clone location
+  - export GOPKG_DIR=$GOPATH/src/gopkg.in/throttled
+  - mkdir -p $GOPKG_DIR
+  - mv $TRAVIS_BUILD_DIR $GOPKG_DIR/throttled.v0
+  - cd $GOPKG_DIR/throttled.v0
```
Comment (brandur) [inline .travis.yml]: I'm probably missing something important here, but does this still run the tests? It looks like the idea is to cover tests for both versions?

**Comment by metcalf** [comment]
@zenazn I'm using interface upgrades here, [inspired by your blog post](http://avtok.com/2014/11/05/interface-upgrades.html). Since at the end of your post you say to "not write your own", I'm curious to hear your take.

Basically there's a simple `LimitResult` interface that can return whether a request should be allowed. For rate limiting in particular, it's useful to be able to expose more information to the client (e.g. an `X-RateLimit-Remaining: 123` header). I implemented that by returning a `LimitResult` that can be optionally upgraded to a `RateLimitResult`. You can see this in action in the [HTTPLimiter](https://github.com/metcalf/throttled/blob/dfe9db940b0993148eed6807a9eef955eba1e692/http.go#L86) and see [docs on the proposed interfaces](https://godoc.org/gopkg.in/metcalf/throttled.v2).

If you have a sec to look, I'd be curious what you think!

**Comment by zenazn** [comment]
I think the most important thing is to document the potential upgrade on both `LimitResult` as well as `RateLimitResult`, e.g., "If the concrete type of `LimitResult` also implements `RateLimitResult`, you will go to space today". Users should also be able to read the documentation to determine which `LimitResult`'s returned by your library also implement `RateLimitResult` so they can plan accordingly.

File: .travis.yml (hunk 1 above)
Comment (metcalf) [inline .travis.yml]: If it finds a Makefile, Travis runs `make` as the test script by default. That's not obvious so I've updated the configuration to be explicit.

Comment (brandur) [inline .travis.yml]: Ah, got it. Thanks.

**Comment by metcalf** [comment]
@zenazn and I talked a bit about the interface yesterday and he made a some good suggestions:
* I'm confused by the number of layers of indirection between NewGCRARateLimiter and a boolean. It seems like LimitResult could plausibly be folded into Limiter.Limit.
* The names "LimitResult" and "RateLimitResult" are equivalent in my mind. Perhaps "Result" and "ResultMetadata" or similar? ("LimitDescription"?)
* Methods named "Reset" typically reset things. "ResetAfter" maybe?

I suggested replacing the indirection with something like the following where the caller could try making assertions on the returned interface:

```
type Limiter interface {
    Limit(key string) (bool, interface{}, error)
}
```

He advised to think in terms of "emit structs, consume interfaces." Which I think would look something like `NewGCRARateLimiter` returning a `GCRARateLimiter` that implements two interfaces:

```
type Limiter interface {
    Limit(key string) (bool, error)
}

type RateLimiter interface {
    RateLimit(key string) (bool, RateLimitContext, error)
}

type RateLimitContext struct {
    Limit, Remaining int
    ResetAfter, RetryAfter time.Duration
}

func NewGCRARateLimiter(store.Store, RateQuota) (GCRARateLimiter, error){ ... }
````

Consumers like the `HTTPLimiter` would accept a `Limiter` and internally attempt to upgrade that type to a `RateLimiter` rather than attempting the upgrade on the return value. 

I like that design but I'm still thinking it through a bit. What do you think @brandur? (and most welcome to keep chiming in @zenazn!)

**Comment by brandur** [comment]
> What do you think @brandur?

Totally up to you @metcalf. I personally wouldn't say that in the situation where we just always returned the result as `(bool, RateLimitContext, error)` that it would be too onerous for consumers that didn't care about the metadata to just ignore it:

``` go
limited, _, err := RateLimit(key)
```

Which saves you from getting into into interface upgrades (which if I'm reading right, seem to mostly be a technique by the Go core team to cheat a bit into coaxing out some extra performance optimization while only exposing nice public interfaces). I have no string opinion though :)

**Comment by metcalf** [comment]
When I was playing with implementing `Limit(key string, count int)` vs a separate `LimitWithCount(key string, count int)` I realized that the "count" concept seems specific to rate limiting. For example, if you were using memory statistics to limit requests (or something silly like limiting every other request), count is no longer relevant. I think that argues in favor of a separate `RateLimiter` interface.

New idea: unless we end up bringing back `memStats` this library only does rate limiting. A totally generic `Limiter` interface is a premature generalization. What about:

```
type RateLimiter interface {
    RateLimit(key string, quantity int) (bool, RateLimitContext, error)
}

type RateLimitContext struct {
    Limit, Remaining int
    ResetAfter, RetryAfter time.Duration
}

func NewGCRARateLimiter(store.Store, RateQuota) (GCRARateLimiter, error){ ... } // Implements `RateLimiter`

// HTTPLimiter faciliates using a RateLimiter to limit HTTP requests.
type HTTPRateLimiter struct {
	DeniedHandler http.Handler
	Error func(w http.ResponseWriter, r *http.Request, err error)
	RateLimiter RateLimiter
	VaryBy interface {
		Key(*http.Request) string
	}
}
```

Arguably this makes the `HTTPRateLimiter` more single-purpose but it's a pretty trivial convenience method if you're not even using it to set rate limiting headers. It's somewhat telling that neither of us do rate limiting in middleware in production -- we call the interface directly.

Curious what you think. I'll take a stab at that implementation so we have something to look at.

**Comment by metcalf** [comment]
OK, another fun question. Currently, a `RateQuota` is defined in terms of a number of requests (`Count`) per time period (`Period`). This can be a bit confusing since it actually permits an instantaneous burst of `Count` requests followed by a sustained rate of `Count/Period` requests. So `PerMin(60)` actually allows 120 requests in the first minute followed by 60 requests per minute after that whereas `PerSec(1)` allows exactly 1 request per second all the time.

Would it be less confusing to define the `RateQuota` in terms of a `MaxBurst` and `Frequency`? `MaxBurst` represents the maximum number of requests in an instant burst. `Frequency` represents the time between requests at a sustained rate (the inverse of `Count/Period`). In our previous example, the `Frequency` is always one second. `MaxBurst` feels intuitive but for some reason `Frequency` doesn't feel as intuitive as `Count/Period` to me. 

Thoughts?

**Comment by brandur** [comment]
@metcalf:

> New idea: unless we end up bringing back memStats this library only does rate limiting. A totally generic Limiter interface is a premature generalization. What about:

+1. [Do One Thing and Do It Well](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well).

I also like that with the `RateLimiter` naming, we're not really painting ourselves into a corner. We could always expand the library if need be (but I doubt we'll need to do this).

> Would it be less confusing to define the RateQuota in terms of a MaxBurst and Frequency? MaxBurst represents the maximum number of requests in an instant burst. Frequency represents the time between requests at a sustained rate (the inverse of Count/Period). In our previous example, the Frequency is always one second. MaxBurst feels intuitive but for some reason Frequency doesn't feel as intuitive as Count/Period to me. 

Oof, good call bringing this up. Okay, I might have to draw this one out to make sure that I've fully wrapped my head around it, but is it possible that we're missing a dimension of configuration here? I found myself going back to the literature a bit to try and get a better grasp for the original algorithm. From Wikipedia:

> “The continuous-state leaky bucket can be viewed as a finite capacity bucket whose real-valued content drains out at a continuous rate of 1 unit of content per time unit and whose content is increased by the increment T for each conforming cell... If at a cell arrival the content of the bucket is less than or equal to the limit value τ, then the cell is conforming; otherwise, the cell is non-conforming. The capacity of the bucket (the upper bound of the counter) is (T + τ)”

Note that at no point does it imply that the limit τ is in derived from T (which is essentially what we're doing by calculating the delay variation tolerance from the defined rate). Should we have a rate _and_ a bucket size? i.e. Something like this?

``` go
type Quota struct {
	// Period between new tokens being added to the bucket.
	FillPeriod time.Duration

	// Maximum number of tokens in the bucket.
	MaxSize int
}
```

You could then turn helpers like `PerMin` into functions to help calculate a period (a little like their `Delay()` implementation already does):

``` go
func PerMin(val int) time.Duration {
	return time.Duration(1.0 / float64(val) * float64(time.Minute))
}
```

You could then alter your configuration to allow different maximum bucket sizes:

``` go
// seed at 60 tokens, fill at 30 tokens a minute
q1 := &Quota{
	FillPeriod: PerMin(30), // the result is FillPeriod = 2 seconds
	MaxSize:    60,
}

// seed at 1 token, fill at 30 tokens a minute
q2 := &Quota{
	FillPeriod: PerMin(30), // the result is FillPeriod = 2 seconds
	MaxSize:    1,
}
```

Note that I pulled the justification from the article on GCRA, but I think that a maximum size would apply to any token bucket-like implementation.

Unless I'm misunderstanding something (likely), I'd probably avoid the term "frequency" just so it's not conflated with the technical definition (number of occurrences per period). So with all of that said ... I'm probably way overcomplicating this, so take it with a grain of salt :)

**Comment by metcalf** [comment]
I think you need two independent parameters but that you can define them in many ways. I think all of the things we've proposed so far are actually equivalent. 

My original proposal (Count and Period) can be confusing because it isn't totally clear that Count represents a burst size. Thus you could accidentally configure things to allow more requests than desired.

Your proposal (FillRate and MaxSize) avoids the burst confusion. I could see someone mistakenly setting MaxSize to something very small (e.g. zero or one). Also having PerMin(30)  and a MaxSize of 1 introduces 3 variables (30, a minute and 1) where we only need 2. Nonetheless, I think your approach is more intuitive.

My final concern with that approach is that it breaks backwards compatibility for the Per* functions. I suspect that a decent number of people are using those to configure their rate limits. Can you think of a way to do it without breaking compatibility? The approach I can think of is a bit janky:

```
type PerSec int

func (p PerSec) Nanoseconds() int64 {
    return time.Second / p
}

func (p PerSec) quota() RateQuota {
    return RateQuota{time.Second / p, p}
}

type RateQuota struct {
    Period interface {
        Nanoseconds() int64
    }
    MaxBurst int
}
```

Basically you rely on the `Nanoseconds` part of the `time.Duration` interface. Ugly, not very self documenting, but compatible.

**Comment by brandur** [comment]
> Your proposal (FillRate and MaxSize) avoids the burst confusion. I could see someone mistakenly setting MaxSize to something very small (e.g. zero or one). Also having PerMin(30) and a MaxSize of 1 introduces 3 variables (30, a minute and 1) where we only need 2. Nonetheless, I think your approach is more intuitive.

Cool! Keep in mind that I was mostly just pitching alternatives and my naming scheme could be terrible. Let's stick to what you've laid out :) Apologies for the equivalent approach too. I realize now that I was somehow looking at an outdated branch when I wrote that up.

Regarding the three variables though: my intention was that there would only be two entities here:

1. A maximum bucket size.
2. A period of fill, probably stored as the time for one token to be added to the bucket.

No. 2 could be deconstructed as two separate variables (and may be internally), but I would hope that this would be for convenience alone; generally the helpers that provide easy mental abstractions like `PerMin` will be leveraged by users to map to whatever representation the internals expect. (So in your example above, 1 per minute just maps to period = 60.)

> My final concern with that approach is that it breaks backwards compatibility for the Per* functions. I suspect that a decent number of people are using those to configure their rate limits. Can you think of a way to do it without breaking compatibility?

Just to drill into this a little more, it seems to me that the primary use of the `Per*` functions would have been as helpers that plug directly into the old rate limiting interface:

``` go
th := throttled.RateLimit(throttled.PerMin(30), &throttled.VaryBy{RemoteAddr: true}, store.NewMemStore(1000))
```

Seeing as any code that looks like this going to have to be rewritten to comply to the V2 interface anyway, I wonder if maybe the breakage is okay? It seems unlikely that anyone is using the helpers independently. i.e.

``` go
q := throttled.PerMin(30)

// calculate my delay!
delay := q.Delay()
```

(Although I suppose it is within the realm of possibility.)

**Comment by metcalf** [comment]
I was actually hoping to maintain compatibility for that case:
https://github.com/throttled/throttled/pull/9/files#diff-4c58ba8920b2c6a341d31c94ab6d3acaR20

I think having the fill period defined as a `time.Duration` might be confusing to people anyway. "X requests per minute" is intuitive but actually the inverse of the period. What about trying to avoid that confusion and maintain compatibility with a `Rate` type:

```
type Rate struct {
    // note that these are private
    count int // ignored except by the compatibility shim
    period time.Duration
}

type RateQuota {
    MaxRate Rate
    MaxBurst int
}

func PerMin(int) Rate { return Rate {count, time.Minute / count} }
...
```

The compatibility shim uses `count` to set `MaxBurst` but otherwise we ignore it. I hate that forcing people to use the `Per*` helpers treats users as stupid but I think it's harmless in practice.

**Comment by brandur** [comment]
> I was actually hoping to maintain compatibility for that case:
> https://github.com/throttled/throttled/pull/9/files#diff-4c58ba8920b2c6a341d31c94ab6d3acaR20

Fair enough! Your design here does look really solid and there's no point breaking compatibility without very good reason. +1 from me.

**Comment by metcalf** [comment]
OK, I think I've implemented all of the changes we've discussed and spent plenty of time agonizing over interfaces. I also plugged this into a few services at Stripe to be sure it would fit (though I haven't merged the branches).

@brandur, I'd love a review of the code whenever you have a chance to wade in. The `varyby*` files haven't changed at all but everything else is sufficiently different that it's probably easier to review from a clean slate than by looking at the diffs.

Meanwhile, I'm planning to branch master to `release-1.x`, update the README on that branch to use `gopkg.in` and tag a `v1.0.0` release. I'm not married to any of those names so happy to change.

**Comment by brandur** [comment]
> OK, I think I've implemented all of the changes we've discussed and spent plenty of time agonizing over interfaces. I also plugged this into a few services at Stripe to be sure it would fit (though I haven't merged the branches).

Awesome! :) I'm going to try adding it to one of my prototype services here as well.

> @brandur, I'd love a review of the code whenever you have a chance to wade in. The varyby* files haven't changed at all but everything else is sufficiently different that it's probably easier to review from a clean slate than by looking at the diffs.

I'll try to take a complete look tonight.

> Meanwhile, I'm planning to branch master to release-1.x, update the README on that branch to use gopkg.in and tag a v1.0.0 release. I'm not married to any of those names so happy to change.

That looks fine to me.

File: deprecated.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
+package throttled
```
Comment (brandur) [inline deprecated.go]: I really like how all of `deprecated` turned out. It's a nice way of organizing all the code that's no longer centrally relevant, and is well-documented in all the right places. I read through and had no questions about it at all.

File: doc.go [hunk 3]
```diff
@@ -1,77 +1,3 @@
...
-//
-// Finally, many examples are provided in the /examples sub-folder of the repository.
-//
-// License
-//
-// The BSD 3-clause license. Copyright (c) 2014 Martin Angers and Contributors.
-//    http://opensource.org/licenses/BSD-3-Clause
-//
-package throttled // import "github.com/throttled/throttled"
+// Package throttled implements various strategies for limiting access to resources,
```
Comment (brandur) [inline doc.go]: Not a big deal at all, but we should consider just making this the same as the description in the `README` (they're close, but worded slightly differently). After this lands, let's also update the GH repo description.

**Comment by brandur** [comment]
Fairly minor, but I find it's often useful to include instructions in the `README` on how to run the project's tests. This helps accelerate new contribution to the project.

Related: is it a common gopkg convention to use its paths in your test suite? It's a little unfortunate because you can't just clone a target repo and run `make` to get the test suite going.

File: rate.go [hunk 4]
```diff
@@ -1,116 +1,237 @@
...
-// The rate limit throttler sets the following headers on the response:
-//
-//    X-RateLimit-Limit : quota
-//    X-RateLimit-Remaining : number of requests remaining in the current window
-//    X-RateLimit-Reset : seconds before a new window
-//
-// Additionally, if the request was denied access, the following header is added:
-//
-//    Retry-After : seconds before the caller should retry
+// A RateLimiter manages limiting the rate of actions by key
```
Comment (brandur) [inline rate.go]: Crazy minor, but you might want to just add punctuation here since you seem to have it everywhere else.

File: rate.go [hunk 5]
```diff
@@ -1,116 +1,237 @@
...
+type RateLimiter interface {
+	// RateLimit checks whether a particular key has exceeded a rate
+	// limit. It also returns a RateLimitContext to provide additional
+	// information about the state of the RateLimiter.
+	//
+	// If the rate limit has not been exceeded, the underlying storage
+	// is updated by the supplied quantity. For example, a quantity of
+	// 1 might be used to rate limit a single request while a greater
+	// quantity could rate limit based on the size of a file upload in
+	// megabytes. If quantity is 0, no update is performed allowing
```
Comment (brandur) [inline rate.go]: Great idea providing a common example for `quantity` here. It might not be too obvious otherwise.

File: http.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	// limiter.  If it is nil, all requests use an empty string key.
+	VaryBy interface {
+		Key(*http.Request) string
+	}
+}
+
+// RateLimit wraps an http.Handler to limit incoming requests.
+// Requests that are not limited will be passed to the handler
+// unchanged.  Limited requests will be passed to the DeniedHandler.
+// If the Limiter returns a RateLimitResult `X-RateLimit-Limit`,
```
Comment (brandur) [inline http.go]: It looks like this comment might be outdated in that it likely describes an interface upgrade.

File: rate.go [hunk 7]
```diff
@@ -1,116 +1,237 @@
...
+	// you to "peek" at the state of the RateLimiter for a given key.
+	RateLimit(key string, quantity int) (bool, RateLimitContext, error)
+}
+
+// RateLimitContext represents the state of the RateLimiter for a
+// given key at the time of the query. This state can be used, for
+// example, to communicate information to the client via HTTP
+// headers. Negative values indicate that the attribute is not
+// relevant to the implementation or state.
+type RateLimitContext struct {
```
Comment (brandur) [inline rate.go]: STOP, BIKESHED TIME. I'm not sure that "context" is the most appropriate name here. Is there a reason that you decided to move away from `RateLimitResult` as described during the interface upgrade pitch?

File: rate.go [hunk 8]
```diff
@@ -1,116 +1,237 @@
...
+	// state. For example, if a rate limiter allows 10 requests per
+	// second and has already received 6 requests for this key this
+	// second, Remaining would be 4.
+	Remaining int
+
+	// ResetAfter is the time until the RateLimiter returns to its
+	// initial state for a given key. For example, if a rate limiter
+	// manages requests per second and received one request 200ms ago,
+	// Reset would return 800ms. You can also think of this as the time
+	// until Limit and Remaining will be equal.
```
Comment (brandur) [inline rate.go]: Excellent descriptions here. Examples FTW.

File: rate.go [hunk 9]
```diff
@@ -1,116 +1,237 @@
...
+	reset, retryAfter time.Duration
+}
+
+func (r *rateLimitResult) Limit() int                { return r.limit }
+func (r *rateLimitResult) Remaining() int            { return r.remaining }
+func (r *rateLimitResult) Reset() time.Duration      { return r.reset }
+func (r *rateLimitResult) RetryAfter() time.Duration { return r.retryAfter }
+
+type Rate struct {
+	period time.Duration //  Time between equally spaced requests at the rate
```
Comment (brandur) [inline rate.go]: So minor it's microscopic: extra whitespace between // and "Time".

File: rate.go [hunk 10]
```diff
@@ -1,116 +1,237 @@
...
+	"fmt"
 	"time"
 )
 
-// Static check to ensure that rateLimiter implements Limiter.
-var _ Limiter = (*rateLimiter)(nil)
+const (
+	// Maximum number of times to retry SetIfNotExists/CompareAndSwap operations
+	// before returning an error.
+	maxCASAttempts = 10
```
Comment (brandur) [inline rate.go]: Thanks for documenting `maxCASAttempts` this time around :) I have a feeling that this will help somebody at some point.

File: rate.go [hunk 11]
```diff
@@ -1,116 +1,237 @@
...
+			}
+			ttl = tat.Sub(now)
+			limited = true
+			break
+		}
+
+		ttl = newTat.Sub(now)
+
+		if tatVal == -1 {
+			updated, err = g.store.SetIfNotExistsWithTTL(key, newTat.UnixNano(), ttl)
```
Comment (brandur) [inline rate.go]: It looks like you [accounted for some of my original feedback on this one](https://github.com/throttled/throttled/pull/8#discussion_r35445148). +1. I have a hard time imagining how a race condition would occur here at this point.

File: store/mem.go [hunk 12]
```diff
@@ -2,89 +2,129 @@ package store
...
+// true. Otherwise, it returns false. If the key does not exist in the
+// store, it returns false with no error. It ignores the ttl.
+func (ms *MemStore) CompareAndSwapWithTTL(key string, old, new int64, _ time.Duration) (bool, error) {
+	valP, ok := ms.get(key, false)
+
+	if !ok {
+		return false, nil
+	}
+
+	return atomic.CompareAndSwapInt64(valP, old, new), nil
```
Comment (brandur) [inline store/mem.go]: Wow. TIL that `atomic` is a thing.

**Comment by brandur** [comment]
@metcalf Hey! Out of a misplaced sense of obligation, I added some really vacuous feedback here, but generally found very little to complain about. Overall this is beautifully written, which is a breath of fresh air given that I rarely see code of this caliber in industry.

I think the biggest thing that I'd make sure to have a story for is easy contribution in that it's clear how the repository is supposed to be cloned and the tests run. I think it's largely ignorance on my part, but it's not clear to me how I'm supposed to clone a gopkg.in link and still make a change and send a patch back to the canonical repository at throttled/throttled.

One thing I realized while reading this is while the current interface has support for removing more than the standard amount of tokens from a bucket, it doesn't have a story for removing fewer. We have a case at work where certain customers have above normal API rate limits which currently looks like a faster drip in the buckets. Because this isn't really possible with GCRA, one option would be to remove a fraction of a single token which of course wouldn't be possible with an `int`. Another option might be to force a baseline of a single token and give a standard account a high multiplier (say `100`) so that you have some room to work with in handing out more generous limits for certain accounts. I don't think that it's worth stopping the line over this, but I'll need to come up with a solution for it before I can switch over some of our main rate limits at work.

Anyway, once again, nice work. Thanks for putting it all together.

**Comment by brandur** [comment]
Hey @metcalf, regarding my comment above as to the possibility of a smaller-than-usual increment rather than one which is larger than usual, what do you think the ramifications would be of a floating point `quantity`?

Here's a patch that is mostly interoperable with what you have already, but adds that idea:

``` patch
diff --git a/rate.go b/rate.go
index 0d22040..e906829 100644
--- a/rate.go
+++ b/rate.go
@@ -23,7 +23,7 @@ type RateLimiter interface {
        // quantity could rate limit based on the size of a file upload in
        // megabytes. If quantity is 0, no update is performed allowing
        // you to "peek" at the state of the RateLimiter for a given key.
-       RateLimit(key string, quantity int) (bool, RateLimitContext, error)
+       RateLimit(key string, quantity float64) (bool, RateLimitContext, error)
 }

 // RateLimitContext represents the state of the RateLimiter for a
@@ -160,7 +160,7 @@ func NewGCRARateLimiter(st GCRAStore, quota RateQuota) (*GCRARateLimiter, error)
 // quantity could rate limit based on the size of a file upload in
 // megabytes. If quantity is 0, no update is performed allowing you
 // to "peek" at the state of the RateLimiter for a given key.
-func (g *GCRARateLimiter) RateLimit(key string, quantity int) (bool, RateLimitContext, error) {
+func (g *GCRARateLimiter) RateLimit(key string, quantity float64) (bool, RateLimitContext, error) {
        var tat, newTat, now time.Time
        var ttl time.Duration
        rlc := RateLimitContext{Limit: g.limit, RetryAfter: -1}
@@ -185,7 +185,7 @@ func (g *GCRARateLimiter) RateLimit(key string, quantity int) (bool, RateLimitCo
                        tat = time.Unix(0, tatVal)
                }

-               increment := time.Duration(quantity) * g.emissionInterval
+               increment := time.Duration(float64(g.emissionInterval) * quantity)
                if now.After(tat) {
                        newTat = now.Add(increment)
                } else {
```

File: store/mem.go (hunk 12 above)
Comment (metcalf) [inline store/mem.go]: Admittedly the godocs basically say "don't use this unless you're really sure you know what you're doing". I could use the mutex here instead if you'd prefer the clarity.

**Comment by metcalf** [comment]
@brandur, I've responded to your comments, thanks for the thorough review! The most substantial thing to look at in the updated diff is the new "Contributing" section of the README. I still find contributing to Go projects confusing...

Regarding the issue of having different rate limits per user, do they occupy different key spaces (e.g. keyed by their account ID)? If so, it might be even clearer to instantiate multiple `RateLimiter` instances that share the same store but have different limits. That way it's clear from just reading the code exactly what limits are being applied to what category of account. If the store is shared, the overhead would be pretty low. Something like:

```
st := store.NewRedisStore(...)
normalLimiter, err := throttled.NewGCRARateLimiter(st, throttled.RateQuota{throttled.PerMin(60), 30})
highValueLimiter, err := throttled.NewGCRARateLimiter(st, throttled.RateQuota{throttled.PerMin(120), 60})
```

If you have a lot of values or want to configure them at runtime you could probably generate a `map[string]RateQuota` that you use to generate a `map[string]RateLimiter`. Would this work / be clearer for your case? If not, how are you defining the rate limits today (e.g. in code, config, database)?

**Comment by metcalf** [comment]
I cut the 1.0.0 release just now. Travis is erroring due to the need for import reshuffling. I'd prefer to just leave it as-is since we're so close to merging this and fixing the build that way. If you think it's worth fixing ASAP, I can backport some of the changes.

File: store/mem.go (hunk 12 above)
Comment (brandur) [inline store/mem.go]: Nah, this is fine as is :)

**Comment by brandur** [comment]
@metcalf 

> I've responded to your comments, thanks for the thorough review! The most substantial thing to look at in the updated diff is the new "Contributing" section of the README. I still find contributing to Go projects confusing...

Hah, you and me both. The new section looks good. Thanks for adding it in!

> Regarding the issue of having different rate limits per user, do they occupy different key spaces (e.g. keyed by their account ID)? If so, it might be even clearer to instantiate multiple RateLimiter instances that share the same store but have different limits. That way it's clear from just reading the code exactly what limits are being applied to what category of account. If the store is shared, the overhead would be pretty low. Something like:

Yeah, +1. Now that you mention it and after thinking about it in a little depth, I realize that what we're currently doing with a multiplier is a weird internal thing that was initially seeded with a very mediocre idea and which over time I got used to and internalized.

We currently pull the multipliers out of a database and they're quite variable so the easiest thing might be to just re-instantiate a limiter for every request. This isn't great, but actually isn't a terrible amount of overhead given that we can re-use the store between them with no problem. Over time it can probably be refactored into normalized tiers of different limits.

tl;dr Let's leave the interface as is. Thanks for acting as a sounding board here.

:ship:

**Comment by metcalf** [comment]
Hrm, so I'm trying to get this deployed to a single service quickly to validate it in production before merging. One annoying thing that came up is that in order to use the in-memory store, you need to have `redigo` because otherwise you can't actually import the `store` package. Should the different concrete store implementations live as sub-packages under `store` (much as `sql` is a subpackage of the empty `database` package in stdlib)?

**Comment by brandur** [comment]
> Hrm, so I'm trying to get this deployed to a single service quickly to validate it in production before merging. One annoying thing that came up is that in order to use the in-memory store, you need to have redigo because otherwise you can't actually import the store package.

Oh wow, great catch!

> Should the different concrete store implementations live as sub-packages under store (much as sql is a subpackage of the empty database package in stdlib)? 

I think that's a great idea.

**Comment by metcalf** [comment]
OK, try 0a2359d on for size. It definitely removes the forced redigo dependency.
* I'm very uncertain about the naming here. Shed color suggestions welcome
* This gave me an opportunity to export the testing helpers publicly so that others can use them to implement custom stores.
* I added the option to return errors from `New` for both stores. Seems like a good thing to have even though we don't need it right now for the Redis store.

**Comment by brandur** [comment]
@metcalf Nice patch!

> * I'm very uncertain about the naming here. Shed color suggestions welcome

Okay, I'm going to bite a little bit on this one :)

I've found generally that when a package name is short and/or ambiguous enough, it's generally a good idea to quality it a little bit. In my opinion, this is at least somewhat relevant for both `mem` and `redis`. I might suggest going with the naming that you used in store's deprecated.go: `memstore` and `redisstore`, both of which adds enough padding to make the names relatively unique.

There is some precendent of this type of pattern in Go's standard library (which I hate to cite as some kind of holy source, but it's somewhat relevant here) in the form of `http/httputil`, `http/httptest`, and `net/rpc/jsonrpc` (which might otherwise be named ambiguously `http/util`, `http/test`, and `net/rpc/json`).

> * This gave me an opportunity to export the testing helpers publicly so that others can use them to implement custom stores.
> * I added the option to return errors from New for both stores. Seems like a good thing to have even though we don't need it right now for the Redis store.

Excellent on both counts!

**Comment by brandur** [comment]
@metcalf +1. :ship:

**Comment by metcalf** [comment]
OK, I just rolled this into production for one service at Stripe -- nothing exploded! I'm going to merge and cut the 2.0.0 release. :tada: happy Friday and thanks for all the help reviewing!

**Comment by brandur** [comment]
@metcalf Woot! Congratulations man.

This is just in time too as I'm going to try and pull it into a service here today :)
